### PR TITLE
Add unit tests in connect-management

### PR DESCRIPTION
**Pull request by Oluane**
## Description

Now that a basic Express mock server is present, the goal of this PR is to add unit tests for this lib "composed" functions. I added in this PR some tests for this two functions: 
* `addIdentityToUser`: 
     * happy path
     * Error throwing when the validation code is unvalid
* `updateIdentityFromUser`: 
    * happy path for primary identity
    * rollback if marking the new address as primary
    * rollback if deleting old primary identity
    * rollback if deleting old non primary identity
    
I also enhanced the basic mock server by adding a router since we now have two different routes. 

Finally, I added  the possibility to customize request's headers during tests, to be able to decide more precisely what operations should respond according to the case we want to test. To do so, I had to completely mock `fetchManagement` just to be able to inject new headers entries in the request while keeping the calls behaviour as identical as if it was a "real" call. I use for that two key: the first, `behaviour`, to determine the behaviour we want to test (for now, there is just `rollback`, but `retry` should come in days to come), the second, `targeted_failure`, stands for the operation on which we want to have a failure.

By writing tests on rollback, I found an issue with our code, where if the removing part of the old identity fails in the end of the flow, in the rollback we tried to remove the old identity again. If the remove part fails, we should rather remove the new one added during the update flow in the rollback. The package will need a new publish once this PR will be merged.

## How Has This Been Tested?

Test suites.

## Types of changes

- [ ] Chore (non-breaking change which refactors / improves the existing code base)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] My change requires a change to a package version.
- [ ] I have updated the `package.json` version accordingly.
- [ ] I have updated the `CHANGELOG.md` version accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by frieda-ewlines** [comment]
Task linked: [CU-9tyhgw [connect-management] - Add unit tests](https://app.clickup.com/t/9tyhgw)

File: management/tests/commands/add-identity-to-user.test.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import { app } from "../mocks/test-server/server";
+
+describe("Add identity to user", () => {
+  let server: Server;
+
+  beforeAll(async () => {
+    await new Promise(
+      (resolve) => (server = app.listen(3000, resolve as () => void)),
+    );
+  });
```
Comment (Fenntasy) [inline management/tests/commands/add-identity-to-user.test.ts:17]:
```suggestion
  beforeAll(async () => {
    await new Promise(
      (resolve) => (server = app.listen(3000, () => resolve())),
    );
  });
```

This should avoid the `as`, doesn't it?

File: management/tests/mocks/test-server/routes/update-identity.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
+import express from "express";
```
Comment (Fuzznimp) [inline management/tests/mocks/test-server/routes/update-identity.ts:1]: I wonder if using a bunch of `if`s instead of `if else` would make the reading better. What do you think?

**Review by Fenntasy — COMMENTED** [review]
I like the way you used headers but I have some questions:
- If we need headers, why couldn't we use an optional parameter on the real `fetchManagement` function? It wouldn't remove the need for mocking the implementation but we could do it with the real function and it would be one less file to care about.
- Do we really need those headers? Couldn't we just have a bigger list of identities that includes the behaviour we want?

**Comment by Oluane** [comment]
>If we need headers, why couldn't we use an optional parameter on the real fetchManagement function? It wouldn't remove the need for mocking the implementation but we could do it with the real function and it would be one less file to care about.

I can see what you mean, and this is another question I came across when working on this. And I chose not to do what you describe here because I didn't wanted to change the real code implementation for the tests to pass. I'm aware that my way of doing this has some drawbacks (like having a copy-pasted file, with the same function, which has just a little change) but I really wanted to have things clearly isolated, whilst trying to keep the global behaviour as close as possible of the real one. 

>Do we really need those headers? Couldn't we just have a bigger list of identities that includes the behaviour we want?

Having a bigger list of identities was the first thing I thought about when I started to write the tests for the rollbacks. Something like a peculiar identity to trigger each case, but I wasn't happy with this idea. Imho, it would complexify both the reading and the understanding of the tests. I kept this way for the primary/non-primary happy path flow because it mimics more closely the real behaviour of `management`, but when it comes to trigger error, I was more confortable to have some kind of metadata to make distinctions rather than relying on the payload we sent in the request. Anyway, I don't know what's really the best way in the end. Wdyt ?

File: management/tests/commands/add-identity-to-user.test.ts (hunk 1 above)
Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts:17]: I will try it 🙂

File: management/tests/mocks/test-server/routes/update-identity.ts (hunk 2 above)
Comment (Oluane) [reply management/tests/mocks/test-server/routes/update-identity.ts:1]: I don't really know what I prefer. One or the other way should work the same, so it's just a matter of reading preferences I guess. I will stand with the team's majority opinion on this one 🙂

Comment (Fuzznimp) [reply management/tests/mocks/test-server/routes/update-identity.ts:1]: Ok, I vote for `if`s then, make this big block breath a bit 😄 .

File: management/tests/mocks/fetch-management.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    fetch,
+  });
+
+  const authLink = setContext((_, { headers }) => {
+    return {
+      headers: {
+        ...headers,
+        authorization: `API_KEY ${managementCredentials.APIKey}`,
+        behaviour: headerValue.behaviour,
+        "targeted-failure": headerValue.targetedFailure,
```
Comment (Meyclem) [inline management/tests/mocks/fetch-management.ts:26]: Rather than adding pretty much the same function that the initial one, what do you think of mocking `execute` instead?
It could still be possible to either return the body of the response, either throw an error.

The advantage being, if we need to change something in the initial function, we wouldn't need to remind ourselves to change the code here directly: if we change code in `fetchManagement` without changing it here, that means some code don't use the real implementation of our code and success in tests but fail in another context

File: management/tests/mocks/test-server/routes/add-identity.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      } else {
+        return response.status(500).json("Something's wrong");
+      }
+    case "addIdentityToUser":
+      if (variables.value === nonPrimaryNewIdentity.value) {
+        return response.status(200).json({
+          data: {
+            addIdentityToUser: {
+              ...nonPrimaryNewIdentity,
+            },
```
Comment (Meyclem) [inline management/tests/mocks/test-server/routes/add-identity.ts:46]:
```suggestion
            addIdentityToUser: nonPrimaryNewIdentity,
```

File: management/tests/mocks/test-server/routes/update-identity.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  switch (operationName) {
+    case "getUserIdentityQuery":
+      if (variables.id === nonPrimaryIdentityToUpdate.id) {
+        return response.status(200).json({
+          data: {
+            provider: {
+              user: {
+                identity: {
+                  ...nonPrimaryIdentityToUpdate,
+                },
```
Comment (Meyclem) [inline management/tests/mocks/test-server/routes/update-identity.ts:25]:
```suggestion
                identity: nonPrimaryIdentityToUpdate,
```

Comment (Meyclem) [reply management/tests/mocks/test-server/routes/update-identity.ts:25]: An the others below

File: management/tests/mocks/test-server/routes/update-identity.ts (hunk 2 above)
Comment (Meyclem) [reply management/tests/mocks/test-server/routes/update-identity.ts:1]: I vote for `if else` because it often avoids some weird things like: 
> I'm going in this `if` because you didn't think about this edge case scenario

**Comment by Meyclem** [comment]
>  I didn't wanted to change the real code implementation for the tests to pass

Totally agree with that part, AAAAAAAAAAmen (You tought me that Fenn :p).
But in the other hand, I agree with @Fenntasy about the two files thing that I don't like either for maintenance purposes.
That's why I proposed to mock the appolo-links `connect` function to behave like we want.

> Anyway, I don't know what's really the best way in the end. Wdyt ?

I would prefer to have hard coded values as less as possible, like the `uuid` that we should randomize.
Having more examples, means more maintenance too when we change something.

File: management/tests/commands/add-identity-to-user.test.ts (hunk 1 above)
Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts:17]:
```typescript
beforeAll(async () => {
    await new Promise<void>(
      (resolve) => (server = app.listen(3000, () => resolve())),
    );
});
```
So I tried with that, and I'm having an issue where the server doesn't seems to close correctly between two tests run. I can't figure out why 😕

Comment (Fenntasy) [reply management/tests/commands/add-identity-to-user.test.ts]: is `server` not assigned properly?

Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts]: What do you mean ?

Comment (Fenntasy) [reply management/tests/commands/add-identity-to-user.test.ts]: Well, it's weird that changing how `beforeAll` works lead to something not closing in the end.
That's why my guess goes to the global variable `server` not being assigned properly.

Which we could test by console logging `server` in the `afterAll` to see if it's undefined or not.

Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts]: Well, I logged `server` in both `afterAll` and they are defined in both case. I still have ` listen EADDRINUSE: address already in use :::3000` when trying to launch the server in the second test running.

Comment (Fenntasy) [reply management/tests/commands/add-identity-to-user.test.ts]: ok let me try your branch

Comment (Fenntasy) [reply management/tests/commands/add-identity-to-user.test.ts]: Ok, I can't explain it at the moment but ...

```diff
   beforeAll(async () => {
     await new Promise<void>(
-      (resolve) => (server = app.listen(3000, () => resolve())),
+      (resolve) => {server = app.listen(3000, () => resolve())},
     );
   });
```

Changing parentheses to curly braces makes it work 😅

Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts]: Alright, something is definitely very strange here. It still doesn't work when trying to run the tests in local, BUT I just pushed this change and it worked in CI. 🤯

Comment (Fenntasy) [reply management/tests/commands/add-identity-to-user.test.ts]: if you still had a server running because of the previous tests, it's logical it failed locally. Maybe after a reboot? 😅

Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts]: But in this case, the first test run using the server won't pass either, no ? Because there, only the one executed in second fails with the already-in-use error locally 🤔

Comment (Fenntasy) [reply management/tests/commands/add-identity-to-user.test.ts]: uuuuh 🤯

File: management/tests/commands/add-identity-to-user.test.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import { nonPrimaryNewIdentity } from "../mocks/identities";
+import { app } from "../mocks/test-server/server";
+
+describe("Add identity to user", () => {
+  let server: Server;
+
+  beforeAll(async () => {
+    await new Promise<void>((resolve) => {
+      server = app.listen(0, () => resolve());
+    });
```
Comment (Oluane) [inline management/tests/commands/add-identity-to-user.test.ts:17]: Giving a port value of 0 here means that it will seek for a random and empty port to launch the server, avoiding us race conditions issues between all tests. Once launched, we can access the port number with : `server.address().port`.

**Comment by Oluane** [comment]
We replaced the whole `fetchManagement` mock by just a mock of the part that does what we want to change in tests, aka the setter function passed to `apollo-link-context`'s`setContext`, in which we can access and modify the request headers. This way, we don't have anymore a copypasted `fetchManagement` test-specific file, and the mocking thing is much more cleaner (Thanks a lot again @Fenntasy 🙏 )

File: management/tests/commands/add-identity-to-user.test.ts (hunk 6 above)
Comment (Fuzznimp) [reply management/tests/commands/add-identity-to-user.test.ts:17]: Sexy af.

**Review by Fuzznimp — APPROVED** [review]
Good for me 👍. Nice one.

**Review by Meyclem — COMMENTED** [review]
I'm not requesting changes because my comments are more proposals here

File: management/src/commands/update-identity-from-user.ts [hunk 7]
```diff
@@ -53,7 +53,7 @@ async function updateIdentityFromUser(
     const identity = {
       userId,
       identityType: getIdentityType(identityToUpdate.type),
-      identityValue: identityToUpdate.value,
+      identityValue,
```
[the remark this answers is not in the document]
Comment (Oluane) [reply management/src/commands/update-identity-from-user.ts:56]: Nope, because in the first place, we want to delete the "old" identity, and not the new one. If this step fails, then we want to delete the "new" identity we just added to rollback in initial state.
Here `identityValue` is the value of the "new" identity we want to replace the old with.

File: management/tests/commands/add-identity-to-user.test.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
+import { Server } from "http";
+import { AddressInfo } from "net";
+
+import { addIdentityToUser } from "../../src/commands";
+import { InvalidValidationCodeError } from "../../src/errors";
+import * as checkVerificationCode from "../../src/queries/check-verification-code";
```
[the remark this answers is not in the document]
Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts:6]: I think I could do that indeed 🙂

File: management/tests/commands/add-identity-to-user.test.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    checkVerificationCode,
+    "checkVerificationCode",
+  );
+
+  test("happy path", async () => {
+    expect.assertions(2);
+
+    const mockedAddIdentityManagementCredentials = {
+      URI: `http://localhost:${
+        (server.address() as AddressInfo).port
```
[the remark this answers is not in the document]
Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts:40]: Nope, if we declare that inside the `beforeAll`, we have a scope issue. 😕

[the remark this answers is not in the document]
Comment (Meyclem) [reply management/tests/commands/add-identity-to-user.test.ts:40]: Not if you declare the variable as a `let` as we do here: https://github.com/fewlinesco/connect-profile/blob/main/test/end-user-api/profile/get-profile.test.ts#L40

[the remark this answers is not in the document]
Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts:40]: Yep, I avoid using let as much as possible, but could be worth doing it that way here.

[the remark this answers is not in the document]
Comment (Oluane) [reply management/tests/commands/add-identity-to-user.test.ts]: Done 🙂

**Review by Fuzznimp — APPROVED** [review]
